mozilla/activity-stream
# Bug 1488758 - Part 1. Consolidate ASR prefs into ASRouterPreferences

The primary function of this patch is to:
- remove the `asrouterExperimentEnabled` pref (turning ASRouter on by default)
- consolidate ASR-related settings to access `ASRouterPreferences.jsm` instead of prefs directly. This reduces the need to parse `messageProvider` multiple times and simplifies the calculation of legacy onboarding/snippets rules.

In addition, it also puts the devtools for asrouter behind a new `asrouter.devtoolsEnabled` pref (since ASR is default-on now).

### Existing issues, to be covered in part 2
- Only show ASR snippets if the user pref is on
- Only show ASR snippets if legacy onboarding snippets have is finished
- Convert telemetry to use the new `ASRouterPreferences` module

### Manual testing instructions
- On a new profile, set `browser.newtabpage.activity-stream.asrouter.devtoolsEnabled` to true
- Visit `about:newtab#asrouter`. Ensure it's working and  has CFR/snippets loaded.

#### Onboarding
- Turn on new onboarding by running this in the console: 
```js
Services.prefs.setStringPref(
  "browser.newtabpage.activity-stream.asrouter.messageProviders",
  JSON.stringify([{"id":"onboarding","type":"local","localProvider":"OnboardingMessageProvider","enabled":true,"cohort":1},{"id":"snippets","type":"remote","url":"https://snippets.cdn.mozilla.net/us-west/bundles/bundle_d6d90fb9098ce8b45e60acf601bcb91b68322309.json","updateCycleInMs":14400000,"enabled":true},{"id":"cfr","type":"local","localProvider":"CFRMessageProvider","enabled":true,"cohort":0}])
);
```
- Ensure the little firefox logo in the left corner does NOT appear on `about:newtab` 
- Visit `about:welcome`, click through fxa experience, ensure the new onboarding experience shows up
- In `about:config`, make sure `browser.onboarding.notification.finished` pref is `true`
- Reset the `browser.newtabpage.activity-stream.asrouter.messageProviders` pref
- Ensure that `browser.onboarding.notification.finished` pref is now `false`
- Visit `about:welcome`. Click through the FXA experience. Ensure you do NOT see the new onboarding experience and the little firefox logo shows up.

#### Snippets
- With `browser.newtabpage.activity-stream.asrouter.messageProviders` pref in the default state and asrouter devtools turned on, ensure, that ASR snippets are showing up on new tab pages.
- With a new tab open in the background, turn off ASR snippets by running the following:
```js
Services.prefs.setStringPref(
  "browser.newtabpage.activity-stream.asrouter.messageProviders",
  JSON.stringify([{"id":"onboarding","type":"local","localProvider":"OnboardingMessageProvider","enabled":true,"cohort":1},{"id":"snippets","type":"remote","url":"https://snippets.cdn.mozilla.net/us-west/bundles/bundle_d6d90fb9098ce8b45e60acf601bcb91b68322309.json","updateCycleInMs":14400000,"enabled":false},{"id":"cfr","type":"local","localProvider":"CFRMessageProvider","enabled":true,"cohort":0}])
);
```
- Check the console of existing and new `about:newtab` pages to ensure that `Legacy snippets initialized` was logged. Ensure ASR snippets do NOT show up.
- With a new tab open in the background, reset the `browser.newtabpage.activity-stream.asrouter.messageProviders` pref
- Check the console of the existing new tab to ensure that `Legacy snippets removed` was logged
